src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,294 @@
+            SyntaxNode reference,
+            ImmutableArray<string> newNamespaceParts, 
+            ISyntaxFactsService syntaxFacts, 
+            out SyntaxNode old, 
+            out SyntaxNode @new)
+        {
+            if (!(reference is SimpleNameSyntax nameRef))
+            {
+                old = @new = null;
+                return false;
+            }
+
+            // A few different cases are handled here:
+            //
+            // 1. When the reference is not qualified (i.e. just a simple name), then there's nothing need to be done.
+            //    And both old and new will point to the original reference.
+            //
+            // 2. When the new namespace is not specified, we don't need to change the qualified part of reference.
+            //    Both old and new will point to the qualified reference.
+            //
+            // 3. When the new namespace is "", i.e. we are moving type referenced by name here to global namespace.
+            //    As a result, we need replace qualified reference with the simple name.
+            //
+            // 4. When the namespace is specified and not "", i.e. we are moving referenced type to a different non-global 
+            //    namespace. We need to replace the qualified reference with a new qualified reference (which is qualified 
+            //    with new namespace.)
+
+            old = syntaxFacts.IsRightSideOfQualifiedName(nameRef) ? nameRef.Parent : nameRef;
+
+            // If no namespace is specified, we just find and return the full NameSyntax.
+            if (old == nameRef || newNamespaceParts.IsDefaultOrEmpty)
+            {
+                @new = old;
+            }
+            else
+            {
+                if (newNamespaceParts.Length == 1 && newNamespaceParts[0].Length == 0)
+                {
+                    // If new namespace is "", then name will be declared in global namespace.
+                    // We will replace qualified reference with simple name + global alias.
+                    @new = SyntaxFactory.AliasQualifiedName(
+                        SyntaxFactory.IdentifierName(SyntaxFactory.Token(SyntaxKind.GlobalKeyword)), 
+                        nameRef.WithoutTrivia());
+                }
+                else
+                {
+                    var aliasQualifier = GetAliasQualifierOpt(old);
+                    var qualifiedNamespaceName = CreateNameSyntax(newNamespaceParts, aliasQualifier, newNamespaceParts.Length - 1);
+                    @new = SyntaxFactory.QualifiedName(qualifiedNamespaceName, nameRef.WithoutTrivia());
+                }
+
+                // We might lose some trivia associated with children of `outerMostNode`.  
+                @new = @new.WithTriviaFrom(old);
+            }
+            return true;
+        }
+
+        protected override string EscapeIdentifier(string identifier)
+            => identifier?.EscapeIdentifier();
```

Comment:
when is this passed a null string?

Reply:
it's not

Reply:
ok.  then `?.` should not be necessary :)